hw/ip/otbn/util/shared/instruction_count_range.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Dict, List, Optional, Set, Tuple
+
+from .control_flow import (ControlGraph, ControlLoc, Cycle, Ecall, ImemEnd,
+                           LoopEnd, LoopStart, Ret, program_control_graph,
+                           subroutine_control_graph)
+from .decode import OTBNProgram
+
+
+def _get_insn_count_range(program: OTBNProgram, graph: ControlGraph,
+                          start_pc: int,
+                          stop_at: str) -> Tuple[int, Optional[int]]:
+    '''Return minimum and maximum instruction counts across control paths.
+
+    In the presence of control-flow cycles or loops with a non-constant
+    number of iterations, the maximum number of iterations will be
+    None, because there's no clear maximum number of times the loop/cycle
+    could run.
+
+    The `stop_at` parameter controls the "end" of the control flow path.
+    Acceptable values are:
+      - "loop end": The end of the next loop (`LoopEnd` instance in the graph)
+      - "ecall": `ecall` instruction (`Ecall` instance in the graph)
+      - "ret" : `ret` instruction going back to the current caller (unmatched
+        `Ret` instance in the graph)
+
+    In all cases, the function will raise an error if it encounters one of the
+    other stopping points; for instance, if an `ecall` instruction appears when
+    `stop_at` = "ret", then there will be an error. The function will return
+    the min/max instruction counts across *all control-flow paths* that reach
+    the desired stopping point.
+    '''
+    section, edges = graph.get_entry(start_pc)
+    sec_count = len(section.get_insn_sequence(program))
+
+    # Special case for when we're in a loop; if we have the loop end and one
+    # other edge, this represents the option to either continue the loop or
+    # stop. In this case, we don't want to follow the "stop" branch, but rather
+    # simply return the min/max from ending the loop here.
+    if stop_at == 'loop end' and any([isinstance(e, LoopEnd) for e in edges]):
+        if len(edges) != 2:
+            raise ValueError(
```

Comment:
nit: Should this be an assertion instead of a `ValueError`? I think this comment applies to other usages as well.